### ISSUE TITLE
persistentStorage option does not respect type of saved data 

### ISSUE THREAD

**Issue by rodkiewicz** [Iteo/cached#105]
Hi, first of all, thanks for a great library.
However, when I am trying to read a cached value from persistent storage, during the first launch of the application, the data returned by the cached functions is of the wrong type. 

To reproduce this error, it is crucial that the application is restarted, that is, the cache is retrieved from persistent storage and not from a variable (from a generated .cached.dart file). 

Please let me know if you are accepting contributions from third party developers. I would like to fix this bug. 

```dart
  @Cached(persistentStorage: true)
  Future<List<CarDto>> _getCars({@IgnoreCache(useCacheOnError: true) bool ignoreCache = true, }) async => carApiDataSource.getCars();
``` 
Error:
```
type 'List<dynamic>' is not a subtype of type 'FutureOr<List<CarDto>>'
#0      _CarRepositoryImpl._getCars (package:car_catalog/data/car/repository/car_repostiory_impl.cached.dart:63:18)
<asynchronous suspension>
#1      new TaskEither.tryCatch.<anonymous closure> (package:fpdart/src/task_either.dart:280:30)
<asynchronous suspension>
#2      TaskEither.match.<anonymous closure> (package:fpdart/src/task_either.dart:168:25)
<asynchronous suspension>
#3      CarListCubit.init (package:car_catalog/presentation/page/car_list_page/cubit/car_list_cubit.dart:18:5)
<asynchronous suspension>

``` 

**Workaround**
Please note that the type of return has changed from   `Future<List<CarDto>>` to `Future<List<dynamic>>`
```dart
@Cached(persistentStorage: true)
Future<List<dynamic>> _getCars({@IgnoreCache(useCacheOnError: true) bool ignoreCache = true, }) async => carApiDataSource.getCars();
``` 

Cache storage implementation 
```dart
@injectable
class CacheStorage extends CachedStorage {
  Future<Box<Map<dynamic, dynamic>>> _openBox(String key) => Hive.openBox(key);

  @override
  Future<void> write(String key, Map<dynamic, dynamic> data) async {
    final box = await _openBox(key);
    return box.put(key, data);
  }

  @override
  Future<Map<String, dynamic>> read(String key) async {
    final box = await _openBox(key);
    final Map<dynamic, dynamic> result = box.get(key) ?? {};
    final Map<String, dynamic> convertedValue = result.map((key, v) => MapEntry(key.toString(), v));
    return convertedValue;
  }

  @override
  Future<void> delete(String key) async {
    final box = await _openBox(key);
    return box.delete(key);
  }

  @override
  Future<void> deleteAll() async {
    //todo: implement deleteAll
  }
}

``` 
My env:
` Flutter (Channel stable, 3.13.4, on macOS 14.0 23A344 darwin-arm64, locale en-PL) `
`cached_annotation: 1.6.0`
`build_runner: 2.4.6`
 `cached: 1.6.0`

**Comment by RSC-88** [comment]
Hello, yes we are accepting PRs from third part developers. So if you have solution please create a PR.

**Comment by kubastick** [comment]
I've submitted a draft pull request that includes a failing test illustrating the issue.
The problem is tied to the behaviour of List.cast - [_If this list contains only instances of R, all read operations will work correctly. If any operation tries to read an element that is not an instance of R, the access will throw instead._](https://api.flutter.dev/flutter/dart-core/List/cast.html)
The issue arises when an incorrect type is returned through the read function in PersistentStorageHolder.
It gets passed unnoticed trough whole cached class.
One potential solution, without breaking backward compatibility, is to implement a check during initialization using a type coercer, [like in this comment](https://github.com/dart-lang/language/issues/3056#issuecomment-1543850183). Unfortunately, implementing this solution would require more time than I'm currently willing to spend on this project.

**Comment by kamilraczka** [comment]
Hello,

Sorry for the late response. Many thanks @kubastick for your diagnosis and the tests. It helped us to write a simple value checker that, on opening PersistentStorageHolder, checks if the method return type matches the stored value type. We've used your tests Kuba and added some additional from our side.

#107 will be included in the 1.6.4 version of cached which we're going to ship today, so I am closing the issue.

Thanks for your help!
